packages/target-ethers-v5/src/codegen/index.ts

```diff
@@ -64,11 +65,25 @@ export function codegenContractTypings(contract: Contract) {
     attach(addressOrName: string): this;
     deployed(): Promise<this>;
 
-    on(event: EventFilter | string, listener: Listener): this;
-    once(event: EventFilter | string, listener: Listener): this;
-    addListener(eventName: EventFilter | string, listener: Listener): this;
-    removeAllListeners(eventName: EventFilter | string): this;
-    removeListener(eventName: any, listener: Listener): this;
+    listeners(eventName?: string): Array<Listener>;
+    off(eventName: string, listener: Listener): this;
+    on(eventName: string, listener: Listener): this;
+    once(eventName: string, listener: Listener): this;
+    removeListener(eventName: string, listener: Listener): this;
+    removeAllListeners(eventName?: string): this;
+
+    listeners<T, G>(eventFilter?: TypedEventFilter<T, G>): Array<TypedListener<T, G>>;
+    off<T, G>(eventFilter: TypedEventFilter<T, G>, listener: TypedListener<T, G>): this;
+    on<T, G>(eventFilter: TypedEventFilter<T, G>, listener: TypedListener<T, G>): this;
+    once<T, G>(eventFilter: TypedEventFilter<T, G>, listener: TypedListener<T, G>): this;
+    removeListener<T, G>(eventFilter: TypedEventFilter<T, G>, listener: TypedListener<T, G>): this;
+    removeAllListeners<T, G>(eventFilter: TypedEventFilter<T, G>): this;
```

Comment:
I think the more specific overloads should be first, since TS checks them top-to-bottom IIRC

Reply:
> is it a good idea to skip the whole typed listeners section if a contract has no defined events?

It's a definitely great idea. Since even if these typed listener section lies there useless for contracts without events, I don't think it would affect normal functionality (lmk if it does). Would it be fine if we do that in a separate PR? If not I can do that here.